src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp

```diff
@@ -437,11 +437,39 @@ class MacroAssembler: public Assembler {
     Assembler::INSN(Rd, Rn, Rm, Ra);                                          \
   }
 
-  WRAP(madd) WRAP(msub) WRAP(maddw) WRAP(msubw)
+  WRAP(madd) WRAP(maddw)
   WRAP(smaddl) WRAP(smsubl) WRAP(umaddl) WRAP(umsubl)
 #undef WRAP
 
 
+  inline void msub(Register Rd, Register Rn, Register Rm, Register Ra) {
+    if (VM_Version::supports_a53mac() && Ra != zr)
+      nop();
+    if (VM_Version::is_neoverse_family()) {
+      /* On Neoverse, MSUB uses the same ALU with SDIV.
+       * The combination of MUL/SUB can utilize multiple ALUs,
+       * and is much faster than MSUB. */
+      mul(rscratch1, Rn, Rm);
+      sub(Rd, Ra, rscratch1);
+    } else {
+      Assembler::msub(Rd, Rn, Rm, Ra);
+    }
+  }
+
+  inline void msubw(Register Rd, Register Rn, Register Rm, Register Ra) {
+    if (VM_Version::supports_a53mac() && Ra != zr)
+      nop();
+    if (VM_Version::is_neoverse_family()) {
+      /* On Neoverse, MSUB uses the same ALU with SDIV.
+       * The combination of MUL/SUB can utilize multiple ALUs,
+       * and is much faster than MSUB. */
```

Comment:
Please refine this comment. I suppose this combination can benefit other situiations which multiple instructions grab the M0, not just for MSUB + SDIV.

Reply:
Comments refined in 

https://github.com/openjdk/jdk/pull/19093/commits/bb4178931eca35cfac26bbdce04c22bdc8026805